pkg/controller/contrailcni/contrailcni_controller.go

```diff
@@ -0,0 +1,162 @@
+					},
+				})
+			}
+			return requests
+		}),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileContrailCNI implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileContrailCNI{}
+
+// ReconcileContrailCNI reconciles a ContrailCNI object
+type ReconcileContrailCNI struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	Client      client.Client
+	Scheme      *runtime.Scheme
+	kubernetes  *k8s.Kubernetes
+	ClusterInfo v1alpha1.CNIClusterInfo
+}
+
+// Reconcile reads that state of the cluster for a ContrailCNI object and makes changes based on the state read
+// and what is in the ContrailCNI.Spec
+func (r *ReconcileContrailCNI) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ContrailCNI")
+	instanceType := "contrailcni"
+	instance := &v1alpha1.ContrailCNI{}
+	ctx := context.TODO()
+
+	if err := r.Client.Get(ctx, request.NamespacedName, instance); err != nil && errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if !instance.GetDeletionTimestamp().IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	contrailCNIConfigName := request.Name + "-" + instanceType + "-configuration"
+	if err := r.configMap(contrailCNIConfigName, instanceType, instance).ensureContrailCNIConfigExist(r.ClusterInfo); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	cniDirs := CniDirs{
+		BinariesDirectory: r.ClusterInfo.CNIBinariesDirectory(),
+		DeploymentType:    r.ClusterInfo.DeploymentType(),
+	}
+
+	var nodesListOptions client.MatchingLabels = instance.Spec.CommonConfiguration.NodeSelector
+	var nodes corev1.NodeList
+	if err := r.Client.List(ctx, &nodes, nodesListOptions); err != nil && errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+	jobReplicas := int32(len(nodes.Items))
+	job := GetJob(cniDirs, request.Name, instanceType, &jobReplicas)
+	for idx, container := range job.Spec.Template.Spec.Containers {
+		instanceContainer := utils.GetContainerFromList(container.Name, instance.Spec.ServiceConfiguration.Containers)
+		if instanceContainer != nil {
+			(&job.Spec.Template.Spec.Containers[idx]).Image = instanceContainer.Image
+		}
+	}
+
+	if err := instance.PrepareJob(job, &instance.Spec.CommonConfiguration, request, r.Scheme, r.Client); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	var clusterJob batchv1.Job
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: job.Name, Namespace: job.Namespace}, &clusterJob); err == nil {
+		if clusterJob.Spec.Completions != &jobReplicas {
+			_ = r.Client.Delete(ctx, &batchv1.Job{ObjectMeta: v1.ObjectMeta{Name: job.Name, Namespace: job.Namespace}})
+			return reconcile.Result{RequeueAfter: 5}, nil
+		}
+	} else {
+		if err := r.Client.Create(ctx, job); err != nil {
```

Comment:
This should be done only if err is type of `NotFound`

Reply:
Done